### ISSUE TITLE
Support Form: math challenge doesn't work from some pages

### ISSUE THREAD

**Issue by pdurbin** [IQSS/dataverse#3036]
In #1327 we added a simple math challenge to slow spammers down from filling out the support form but in some places the operands are now missing. For example on https://demo.dataverse.org running v. 4.3 build 22-1e451ff if you sign up, click the "welcome" notification, and click "Dataverse Support", you don't see any number to add together:

![account_-_demo_dataverse_-_2016-03-28_09 01 11](https://cloud.githubusercontent.com/assets/21006/14078442/d8b7e4d6-f4c3-11e5-970f-08c72dcc4eb0.png)

As noted by @kcondon this regression appears in additional places in the Shib-related pull request at #3025 as of 4d332e0:
- 403 pages: https://shibtest.dataverse.org/403.xhtml
- 404 pages: https://shibtest.dataverse.org/404.xhtml
- Login Page when Shib is enabled: https://shibtest.dataverse.org/loginpage.xhtml

Interestingly, if you click the normal "Support" link in the header first, the problem goes away. From a quick look it seems like the problem started happening when we staring using constructions like `<a href="#" onclick="event.preventDefault();PF('contactForm').show();">Dataverse Support</a>` in Bundle.properties. I'm not actually sure what the fix should be. Here's a list of where this construction is used in pull request #3025:

```
murphy:dataverse pdurbin$ grep contactForm src/main/java/Bundle.properties | grep -i onclick | sed G

user.isShibUser=Account information cannot be edited when logged in through an institutional account. Leaving your institution? Please contact <a href="#" onclick="event.preventDefault();PF('contactForm').show();">Dataverse Support</a> for assistance.

notification.welcome=Welcome to {0} Dataverse! Get started by adding or finding data. Have questions? Check out the <a href="{1}/{2}/user/index.html" title="{0} Dataverse User Guide" target="_blank">User Guide</a> or contact <a href="#" onclick="event.preventDefault();PF(&#39;contactForm&#39;).show();" title=\u201dContact {0} Dataverse Support\u201d>Dataverse Support</a> for assistance.

login.institution.support=Leaving your institution? Please contact <a href="#" onclick="event.preventDefault();PF('contactForm').show();">Dataverse Support</a> for assistance.

error.404.message=<strong>Page Not Found</strong> - The page you are looking for was not found. If you believe this is an error, please contact <a href="#" onclick="event.preventDefault();PF('contactForm').show();">Dataverse Support</a> for assistance.

error.403.message=<strong>Not Authorized</strong> - You are not authorized to view this page. If you believe this is an error, please contact <a href="#" onclick="event.preventDefault();PF('contactForm').show();">Dataverse Support</a> for assistance.
```

**Comment by pdurbin** [comment]
@scolapasta thanks for agreeing during this morning's standup to take a look at this issue. Again, the easiest fix is to remove the `onclick ... PF` stuff from https://github.com/IQSS/dataverse/blob/v4.3/src/main/java/Bundle.properties#L129 for example and have it say "Click the link in the header" but maybe you can come up with a better fix. @landreev seems to agree that PrimeFaces code doesn't really belong in the bundle and if that's the case maybe we should put this in our style guide: http://guides.dataverse.org/en/4.3/developers/coding-style.html

@scolapasta during the meeting I said I'd also give two more issues to you that were kicked back to me. Here are links to the comments from @kcondon:

- https://github.com/IQSS/dataverse/issues/2046#issuecomment-202388249
- https://github.com/IQSS/dataverse/issues/2951#issuecomment-202388309

Since the [2939-shib](https://github.com/IQSS/dataverse/tree/2939-shib) branch that pull request #3025 is based on is particularly effected, I would suggest fixing the problem there.

**Comment by scolapasta** [comment]
Giving this to @mheppler. In 4.3 for the notification, and in the shib branch for all the other support references, he added these links to Support.  (403 and 404 used to say, click support link on the header). 

However these references in the bundle only open up the dialog box and don't do what happens when you click on the link in the header, i.e call to the server to initialize the popup. 

So (and @pdurbin mentions some of this above we need to:
a) either go back to just saying "please click link in header"
b) find javascript that will actually click the link as if the user clicked in
c)put the logic on each of the links on the 403, 404, shib login, and account pages to initialize.

The notification is a different case. In particular, we do also send out notification texts in e-mail. So we may not want to have this in the first place, as you're not in the app, i.e. can't open up a link.

**Comment by scolapasta** [comment]
Here is the code from the header:

                        <h:form class="navbar-form navbar-left navbar-form-link">
                           <p:commandLink value="#{bundle['header.support']}" oncomplete="PF('contactForm').show()" update=":contactDialog" actionListener="#{sendFeedbackDialog.initUserInput}">
                                <f:setPropertyActionListener target="#{sendFeedbackDialog.messageSubject}" value=""/>
                                <f:setPropertyActionListener target="#{sendFeedbackDialog.recipient}" value="#{null}"/>
                                <f:setPropertyActionListener target="#{sendFeedbackDialog.userMessage}" value=""/>
                                <f:setPropertyActionListener target="#{sendFeedbackDialog.userEmail}" value=""/>
                           </p:commandLink>
                        </h:form>


In particular, some of the properties would be set differently depending on where you clicked the page from, so fir example, we could prepopulate messageSubject to be "404" if from that page.

This, of course, would suggest that c above would be the course of action.

**Comment by mheppler** [comment]
After reviewing all the comments and looking at the code, I believe that best of the three options outlined above is "c) put the logic on each of the links on the 403, 404, shib login, and account pages to initialize".

Passing this ticket to Gustavo to implement this or better yet find an even more dynamic solution that would allow bundle messages with contact support links to be easily adding going forward, without needing a custom solution.

**Comment by eaquigley** [comment]
+1 to @mheppler's comment. I do not think it should be changed to "click the link in the header". Would not be a good user experience practice.

**Comment by pdurbin** [comment]
When @scolapasta gave this issue to me the other day we discussed it and decided to

- Revert the 403 and 404 pages back to how they are in 4.3 (no links to support) - done in 656a074
- Remove all `PF` code from the bundle - done in adf1e34

In total there are 5 places to check:

- 403 page
- 404 page
- notification on signup
- Login page when Shib is enabled (math question should now work)
- Account page for Shib user (math question should now work)

Ideally, I would have gotten feedback from @eaquigley on how the welcome notification looks but it's a busy day for her so I used my best judgement to fix the regression in 4.3 that was introduced in #2966 (that issue should be tested or retested). Here's a before and after:

## Notification Before

![screen shot 2016-04-19 at 1 35 53 pm](https://cloud.githubusercontent.com/assets/21006/14650013/d1959b6e-0637-11e6-8e25-6070830b3fd2.png)


## Notification After

![screen shot 2016-04-19 at 1 59 35 pm](https://cloud.githubusercontent.com/assets/21006/14650012/d18cdd30-0637-11e6-8378-1462a7484d86.png)

Passing to QA.

**Comment by eaquigley** [comment]
Discussed with @pdurbin and to avoid dead ends with the Support text, he will change the text to say, "or email (insert system email address link) for assistance."

**Comment by pdurbin** [comment]
Change of plan. @kcondon and @eaquigley I got the popup working in 083f279 so we'll go with that (the intended fix for #2966). Passing to QA.

@eaquigley I'd like you to be able to demo the code at https://shibtest.dataverse.org but I'm having trouble getting the IdP at http://www.testshib.org to recognize it. I'll keep working on it.

**Comment by kcondon** [comment]
All place work, closing.

**Comment by pdurbin** [comment]
For the record, since things changed from my earlier update, here is the state of the code:

- reverted to 4.3 behavior
  - 403 page
  - 404 page
- math question should now work
  - notification on signup
  - Login page when Shib is enabled
  - Account page for Shib user

**Comment by scolapasta** [comment]
Added code in widgets branch for 403 and 404 pages to work correctly.

**Comment by scolapasta** [comment]
There is one last place to fix (not listed above) and that's in the messages that come up if you get a caught exception. The example I can think of is: try to assign the same role assignment a 2nd time.

**Comment by kcondon** [comment]
Why was this reopened?

**Comment by kcondon** [comment]
I tested the error pages math challenge and they work now. Passing back to Dev to fix the last problem Gustavo mentioned.

**Comment by pdurbin** [comment]
On dataverse-internal running v. 4.4 build 132-47b04ed I just noticed that if you enter a bad username and password, you don't see the math challenge:

![log_in_-_root_dataverse_-_2016-06-29_13 13 25](https://cloud.githubusercontent.com/assets/21006/16461605/589ccdc6-3dfb-11e6-94c4-6059ee69623f.png)

**Comment by kcondon** [comment]
Two other places this fails is the harvest exception banner mentioned above in the closed ticket, #3265 
and also in the confirm email expired email link landing page.

**Comment by mheppler** [comment]
Since this is a bug that effects the login and sign up workflows, I thought it was important to look into fixing it in association with the ORCID/OAuth2 development in #3338.

I took the latest and greatest "please contact Dataverse Support for assistance" off of the 403 pg and applied it to `messages.xhtml` in order to fix the broken math validation in the login workflow, but was not able to get that working.

```
                    <o:form useRequestURI="true">
                        <h:outputText value="#{bundle['error.403.message']} " escape="false"/>
                        <h:outputFormat value="#{bundle['error.support.message']}" escape="false">
                            <o:param>
                                <p:commandLink value="#{bundle['contact.support']}" oncomplete="PF('contactForm').show()" update=":contactDialog" actionListener="#{sendFeedbackDialog.initUserInput}">
                                    <f:setPropertyActionListener target="#{sendFeedbackDialog.messageSubject}" value=""/>
                                    <f:setPropertyActionListener target="#{sendFeedbackDialog.recipient}" value="#{null}"/>
                                    <f:setPropertyActionListener target="#{sendFeedbackDialog.userMessage}" value=""/>
                                    <f:setPropertyActionListener target="#{sendFeedbackDialog.userEmail}" value=""/>
                                </p:commandLink>                                
                            </o:param>
                        </h:outputFormat>   
                    </o:form>
```
After break, I would like to review this with @scolapasta and @pdurbin to see what we're missing.

**Comment by djbrooke** [comment]
We discussed this in the planning meeting on 1/4. No steps aside from "investigation" were identified. Sometimes that's how it goes. 

Thanks @mheppler for the detailed notes above.

**Comment by pdurbin** [comment]
@scolapasta while you're in this part of the code, can you please take a look at #3545 opened by @pameyer and see if it's an easy fix to incorporate?

**Comment by scolapasta** [comment]
This issue is about the popup and not the actual e-mail that gets sent. Discussed with @djbrooke and we'll keep the scope here limited to this.

**Comment by pdurbin** [comment]
I worked on this issue in the past so I'll do some code review.

**Comment by pdurbin** [comment]
I'm happy with the code in the 3036-support-form-math-bug so I went ahead and created pull request #3573 based on it. I did the following:

- Tested to make sure the math bug is gone when you try to log in with no username as described at https://github.com/IQSS/dataverse/issues/3036#issuecomment-229424569
- Tested to make sure the math bug is gone when you arrive at the http://localhost:8080/confirmemail.xhtml interstitial page and can't proceed due the the token being expired or whatever.
- Updated the the branch with the latest from "develop".

Note that I could not test the math bug scenario mentioned in #3265 because I got a proper error message when I tried to use more than 30 characters for a client nickname.

In short, I believe the math challenge now works everywhere. If there's a place where it doesn't work, I don't know where it is.

Passing to QA.

**Comment by kcondon** [comment]
Tested:
1. Main nav bar
2. 403, 404.xhtml pages
3. Shib enabled log in page
4. Caught exception page (2x role assign)
5. Email verification expire message
6. Log in no user name error message.

This appears to be working. Closing.

**Comment by kcondon** [comment]
@pdurbin Works but needs to have merge conflicts resolved.

**Comment by kcondon** [comment]
All set. Closing.

**Comment by djbrooke** [comment]
This was pulled back into the development column because the fix broke publish. This has been fixed, but a different approach will need to be taken here.

**Comment by djbrooke** [comment]
I'm moving this into the backlog. We'll address this in a future sprint.

**Comment by pdurbin** [comment]
> I'm moving this into the backlog. We'll address this in a future sprint.

This issue goes on and on. Let's create a fresh one when we're ready to estimate it. Closing.

**Comment by kcondon** [comment]
This is still an issue, encountered when testing: https://github.com/IQSS/dataverse/pull/6136

**Comment by pdurbin** [comment]
Closing in favor of #6307
